File: internal/contracts/module.go
package contracts

import (
	"fmt"
	"regexp"
	"strings"
)

var moduleIDPattern = regexp.MustCompile(`^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$`)
var semanticVersionPattern = regexp.MustCompile(`^\d+\.\d+\.\d+$`)

type ModuleDependency struct {
	ID       string `json:"id"`
	Contract string `json:"contract"`
}

type ModuleManifest struct {
	ID           string             `json:"id"`
	Version      string             `json:"version"`
	Layer        string             `json:"layer"`
	Required     bool               `json:"required"`
	Capabilities []string           `json:"capabilities"`
	Dependencies []ModuleDependency `json:"dependencies"`
	ConfigSchema string             `json:"configSchema"`
	DataOwner    string             `json:"dataOwner"`
	HealthCheck  string             `json:"healthCheck"`
}

type ModuleReadModelEntry struct {
	Manifest ModuleManifest `json:"manifest"`
	Status   string         `json:"status"`
}

type CapabilityReadModel struct {
	SchemaVersion   int                           `json:"schemaVersion"`
	PlatformModules []ModuleReadModelEntry        `json:"platformModules"`
	RuntimeModules  []ModuleManifest              `json:"runtimeModules"`
	RuntimeStatus   string                        `json:"runtimeStatus"`
	Engines         map[string]EngineCapabilities `json:"engines"`
}

type EngineCapabilities struct {
	Approval   bool `json:"approval"`
	Resume     bool `json:"resume"`
	Steer      bool `json:"steer"`
	ToolEvents bool `json:"toolEvents"`
	Usage      bool `json:"usage"`
}

func ValidateModuleGraph(manifests []ModuleManifest) error {
	modules := make(map[string]ModuleManifest, len(manifests))
	for _, manifest := range manifests {
		if !moduleIDPattern.MatchString(manifest.ID) || !semanticVersionPattern.MatchString(manifest.Version) || !validModuleLayer(manifest.Layer) || len(manifest.Capabilities) == 0 || strings.TrimSpace(manifest.ConfigSchema) == "" || strings.TrimSpace(manifest.DataOwner) == "" || strings.TrimSpace(manifest.HealthCheck) == "" {
			return fmt.Errorf("invalid_module_manifest:%s", manifest.ID)
		}
		for _, capability := range manifest.Capabilities {
			if strings.TrimSpace(capability) == "" {
				return fmt.Errorf("invalid_module_manifest:%s", manifest.ID)
			}
		}
		if _, exists := modules[manifest.ID]; exists {
			return fmt.Errorf("duplicate_module_id:%s", manifest.ID)
		}
		modules[manifest.ID] = manifest
	}
	for _, manifest := range manifests {
		for _, dependency := range manifest.Dependencies {
			if !moduleIDPattern.MatchString(dependency.ID) || strings.TrimSpace(dependency.Contract) == "" {
				return fmt.Errorf("invalid_module_dependency:%s:%s", manifest.ID, dependency.ID)
			}
			if _, exists := modules[dependency.ID]; !exists {
				return fmt.Errorf("missing_module_dependency:%s:%s", manifest.ID, dependency.ID)
			}
		}
	}
	visiting := make(map[string]bool, len(manifests))
	visited := make(map[string]bool, len(manifests))
	var visit func(string) error
	visit = func(id string) error {
		if visiting[id] {
			return fmt.Errorf("cyclic_module_dependency:%s", id)
		}
		if visited[id] {
			return nil
		}
		visiting[id] = true
		for _, dependency := range modules[id].Dependencies {
			if err := visit(dependency.ID); err != nil {
				return err
			}
		}
		delete(visiting, id)
		visited[id] = true
		return nil
	}
	for id := range modules {
		if err := visit(id); err != nil {
			return err
		}
	}
	return nil
}

func validModuleLayer(layer string) bool {
	switch layer {
	case "web", "platform", "runtime", "adapter":
		return true
	default:
		return false
	}
}

File: packages/contracts/src/module.ts
import { z } from "zod";
import { engineCapabilitiesSchema, engineIdSchema } from "./engine.js";

const moduleIdSchema = z
  .string()
  .regex(/^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/, "invalid module ID");

export const moduleDependencySchema = z.object({
  id: moduleIdSchema,
  contract: z.string().min(1).max(120),
});

export const moduleManifestSchema = z.object({
  id: moduleIdSchema,
  version: z.string().regex(/^\d+\.\d+\.\d+$/, "semantic version required"),
  layer: z.enum(["web", "platform", "runtime", "adapter"]),
  required: z.boolean(),
  capabilities: z.array(z.string().min(1)).min(1),
  dependencies: z.array(moduleDependencySchema),
  dataOwner: z.string().min(1),
  healthCheck: z.string().min(1),
});
export type ModuleManifest = z.infer<typeof moduleManifestSchema>;

export const moduleManifestListSchema = z.array(moduleManifestSchema);

export const moduleHealthSchema = z.enum([
  "healthy",
  "unhealthy",
  "unavailable",
  "unknown",
  "disabled",
]);

export const capabilityReadModelSchema = z.object({
  schemaVersion: z.literal(1),
  platformModules: z.array(
    z.object({ manifest: moduleManifestSchema, status: moduleHealthSchema }),
  ),
  runtimeModules: moduleManifestListSchema,
  runtimeStatus: moduleHealthSchema.exclude(["disabled"]),
  engines: z.partialRecord(engineIdSchema, engineCapabilitiesSchema),
});
export type CapabilityReadModel = z.infer<typeof capabilityReadModelSchema>;

export function validateModuleGraph(
  manifests: readonly ModuleManifest[],
): void {
  const modules = new Map(manifests.map((manifest) => [manifest.id, manifest]));
  if (modules.size !== manifests.length) throw new Error("duplicate_module_id");
  for (const manifest of manifests) {
    for (const dependency of manifest.dependencies) {
      if (!modules.has(dependency.id))
        throw new Error(
          `missing_module_dependency:${manifest.id}:${dependency.id}`,
        );
    }
  }
  const visiting = new Set<string>();
  const visited = new Set<string>();
  const visit = (id: string): void => {
    if (visiting.has(id)) throw new Error(`cyclic_module_dependency:${id}`);
    if (visited.has(id)) return;
    visiting.add(id);
    for (const dependency of modules.get(id)!.dependencies)
      visit(dependency.id);
    visiting.delete(id);
    visited.add(id);
  };
  for (const id of modules.keys()) visit(id);
}
